mujoco_py/mjpid.pyx

```diff
@@ -24,77 +81,191 @@ cdef enum USER_DEFINED_ACTUATOR_PARAMS:
     IDX_DERIVATIVE_GAIN_SMOOTHING = 4,
     IDX_ERROR_DEADBAND = 5,
 
-
 cdef enum USER_DEFINED_CONTROLLER_DATA:
     IDX_INTEGRAL_ERROR = 0,
     IDX_LAST_ERROR = 1,
     IDX_DERIVATIVE_ERROR_LAST = 2,
-    NUM_USER_DATA_PER_ACT = 3,
+    # Needs to be max() of userdata needed for all control modes. 5 needed for cascaded PI
+    NUM_USER_DATA_PER_ACT = 5,
+
+cdef mjtNum c_pid_bias(const mjModel*m, const mjData*d, int id):
+    """
+    To activate PID, set gainprm="Kp Ti Td iClamp errBand iSmooth" in a general type actuator in mujoco xml
+    """
+    cdef double dt_in_sec = m.opt.timestep
+    cdef int NGAIN = int(const.NGAIN)
 
+    result = _pid(parameters=PIDParameters(
+        dt_seconds=dt_in_sec,
+        setpoint=d.ctrl[id],
+        feedback=d.actuator_length[id],
+        Kp=m.actuator_gainprm[id * NGAIN + IDX_PROPORTIONAL_GAIN],
+        error_deadband=m.actuator_gainprm[id * NGAIN + IDX_ERROR_DEADBAND],
+        integral_max_clamp=m.actuator_gainprm[id * NGAIN + IDX_INTEGRAL_MAX_CLAMP],
+        integral_time_const=m.actuator_gainprm[id * NGAIN + IDX_INTEGRAL_TIME_CONSTANT],
+        derivative_gain_smoothing=m.actuator_gainprm[id * NGAIN + IDX_DERIVATIVE_GAIN_SMOOTHING],
+        derivative_time_const=m.actuator_gainprm[id * NGAIN + IDX_DERIVATIVE_TIME_CONSTANT],
+        previous_errors=PIDErrors(
+            error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_LAST_ERROR],
+            derivative_error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_DERIVATIVE_ERROR_LAST],
+            integral_error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_INTEGRAL_ERROR],
+        )))
 
-cdef mjtNum c_zero_gains(const mjModel* m, const mjData* d, int id) with gil:
-    return 0.0
+    d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_LAST_ERROR] = result.errors.error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_DERIVATIVE_ERROR_LAST] = result.errors.derivative_error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_INTEGRAL_ERROR] = result.errors.integral_error
 
+    f = result.output
 
-cdef mjtNum c_pid_bias(const mjModel* m, const mjData* d, int id) with gil:
+    cdef double effort_limit_low = m.actuator_forcerange[id * 2]
+    cdef double effort_limit_high = m.actuator_forcerange[id * 2 + 1]
+    
+    if effort_limit_low != 0.0 or effort_limit_high != 0.0:
+        f = fmax(effort_limit_low, fmin(effort_limit_high, f))
+    return f
+
+cdef enum USER_DEFINED_ACTUATOR_PARAMS_CASCADE:
+    IDX_CAS_PROPORTIONAL_GAIN = 0,
+    IDX_CAS_INTEGRAL_TIME_CONSTANT = 1,
+    IDX_CAS_INTEGRAL_MAX_CLAMP = 2,
+    IDX_CAS_PROPORTIONAL_GAIN_V = 3,
+    IDX_CAS_INTEGRAL_TIME_CONSTANT_V = 4,
+    IDX_CAS_INTEGRAL_MAX_CLAMP_V = 5,
+    IDX_CAS_EMA_SMOOTH_V = 6,
+    IDX_CAS_MAX_VEL = 7,
+
+cdef enum USER_DEFINED_CONTROLLER_DATA_CASCADE:
+    IDX_CAS_INTEGRAL_ERROR = 0,
+    IDX_CAS_INTEGRAL_ERROR_V = 1,
+    IDX_CAS_LAST_ERROR = 2,
+    IDX_CAS_LAST_ERROR_V = 3,
```

Comment:
last_error or last_error_v is used to calculate derivative (D) terms, which is not used in the current cascaded PID controller.

Reply:
yup, just removed in 086f30d, thanks for noticing!